src/ext/handlersFilterKey.c

```diff
@@ -1,163 +1,354 @@
+    structCb->handleDbSize = filterDbSize;
+
+    /* string */
+    structCb->handleString = filterString;
+    /* list */
+    structCb->handleListLP = filterListLP;
+    structCb->handleListZL = filterListZL;
+    structCb->handleListPlain = filterListPlain;
+    /* hash */
+    structCb->handleHashPlain = filterHashPlain;
+    structCb->handleHashZL = filterHashZL;
+    structCb->handleHashLP = filterHashLP;
+    structCb->handleHashZM = filterHashZM;
+
+    /* set */
+    structCb->handleSetPlain = filterSetPlain;
+    structCb->handleHashZM = filterSetIS;
+    structCb->handleSetLP = filterSetLP;
+
+    /* func */
+    structCb->handleFunction = filterFunction;
+    /* module */
+    structCb->handleModule = filterModule;
 }
 
-RdbxFilterKey *RDBX_createHandlersFilterKey(RdbParser *p,
-                                            const char *keyRegex,
-                                            uint32_t flags)
-{
-    RdbxFilterKey *ctx;
-    UNUSED(flags);
+static void defaultFilterRawCb(RdbHandlersRawCallbacks *rawCb) {
+    memset(rawCb, 0, sizeof(*rawCb));
+    /* common */
+    rawCb->handleNewKey = filterNewKey;
+    rawCb->handleEndKey = filterEndKey;
+    rawCb->handleNewDb = filterNewDb;
+    rawCb->handleDbSize = filterDbSize;
 
-    CallbacksUnion callbacks;
-    memset (&callbacks, 0, sizeof(callbacks));
+    //callbacks.rawCb.handleBeginModuleAux  /* not partof keyspace */
+    rawCb->handleBegin = filterRawBegin;
+    rawCb->handleFrag = filterFrag;
+    rawCb->handleEnd = filterRawEnd;
+}
 
-    if ( (ctx = RDB_alloc(p, sizeof(RdbxFilterKey))) == NULL)
+static RdbxFilter *createHandlersFilterCommon(RdbParser *p,
+                                              const char *keyRegex,
+                                              RdbDataType *type,
+                                              int *dbnum,
+                                              uint32_t exclude) {
+    RdbRes (*handleNewKey)(RdbParser *p, void *userData, RdbBulk key, RdbKeyInfo *info) = filterNewKey;
+    RdbRes (*handleNewDb)(RdbParser *p, void *userData,  int dbnum) = filterNewDb;
+    RdbxFilter *ctx;
+
+    if ( (ctx = RDB_alloc(p, sizeof(RdbxFilter))) == NULL)
         return NULL;
 
     ctx->regexInitialized = 0;
 
-    /* compile the regular expression */
-    if (regcomp(&ctx->regex_compiled, keyRegex, REG_EXTENDED) != 0) {
-        RDB_reportError(p, (RdbRes) RDBX_ERR_FILTER_FAILED_COMPILE_REGEX,
-                        "FilterKey: Error compiling regular expression");
-        deleteFilterKeyCtx(p, ctx);
-        return NULL;
-    } else {
+    /* specific if-else init to filter regex/type/dbnum */
+    if (keyRegex) {   /* filter keys by regex */
+        /* compile the regular expression */
+        if (regcomp(&ctx->regex_compiled, keyRegex, REG_EXTENDED) != 0) {
+            RDB_reportError(p, (RdbRes) RDBX_ERR_FILTER_FAILED_COMPILE_REGEX,
+                            "FilterKey: Error compiling regular expression");
```

Comment:
I'm not familiar with this function but looks like regcomp() reports errors as well (via regerror()). Do you think we should log it?

Reply:
Added it to the report. Added test as well.